syntax = "proto3";

package saq.sandwich.proto.api.v1;


import "proto/api/v1/tls.proto";

// Supported protocols.
enum Protocol {
  // Unspecified protocol.
  PROTO_UNSPECIFIED = 0;
  // TLS 1.3
  PROTO_TLS_13 = 1;
}

// Supported implementations.
enum Implementation {
  // Unspecified implementation.
  IMPL_UNSPECIFIED = 0;
  // OpenSSL 1.1.1.
  IMPL_OPENSSL1_1_1 = 1;
  // OpenSSL 1.1.1 with liboqs.
  IMPL_OPENSSL1_1_1_OQS = 2;
}

// Socket options to apply to tunnels.
enum SocketOptions {
  // No-op options.
  // Note: we cannot use an `optional uint32` for `socket_options`, because
  // optional fields are not supported by the protobuf crate (Rust):
  // > is a proto3 file that contains optional fields, but code generator.
  // > protoc-gen-rust hasn't been updated to support optional fields in proto3.
  // > Please ask the owner of this code generator to support proto3 optional.
  SOCK_UNSPECIFIED = 0;

  // Non-blocking mode.
  SOCK_NONBLOCK = 1;
}

// Client options.
message ClientOptions {
  oneof opts {
    // TLS options.
    TLSClientOptions tls = 1;
  }
  // Next id: 2
}

// Server options.
message ServerOptions {
  oneof opts {
    // TLS options.
    TLSServerOptions tls = 1;
  }
  // Next id: 2
}

// A configuration for a sandwich tunnel (server or client).
message Configuration {
  // The protocol to use.
  Protocol protocol = 1;

  // The implementation to use.
  Implementation impl = 2;

  // Bitwise-OR of SocketFlags.
  uint32 socket_options = 3;

  // The option specific to the pair {protocol, impl}.
  oneof opts {
    // Client options.
    ClientOptions client = 4;
    // Server options.
    ServerOptions server = 5;
  }
  // Next id: 7
}
